### DISCUSSION TITLE
Base64 유닛테스트 집단지성으로 모으기!

### DISCUSSION THREAD

**Discussion by YoungSeokHong** [euphony-io/euphony#87] [General]
Base64의 테스트 셋도 다같이 만들어보면 좋을 것 같아서 Dicussion을 만들었습니다.

Base64도 Base32와 원리는 같습니다. 혹시 아직도 진법 변환에 대해서 잘 모르시겠는 분들은 #83 을 참고해주길 부탁드리겠습니다.
Base64와 Base32의 차이점울 중심으로 하여 간략하게 설명하겠습니다.
Base64에서도 Base32에서처럼 뒤에서부터 읽어오는 방식으로 진법변환을 합니다. 이때 Base32에서는 20비트씩 끊어 읽던 것과 다르게 Base64에서는 4와 6의 최소공배수를 사용하여 12개씩 끊어서 읽습니다.

테스트 셋을 만드는 방법이 Base32에 비해서 조금 많이 복잡한데
아쉽게도 인터넷에 있는 Hex to Base64진법 변환기는 앞에서부터 읽고 '='으로 padding을 하는 방식으로, 저희의 Base64와는 다른 계산법을 이용해서  테스트 셋을 만들 때 사용하지 못합니다. ㅠㅠ

## 테스트 셋 만드는 방법
1. 16진수를 2진수로 바꿔준다.

![image](https://user-images.githubusercontent.com/80624082/135959317-96ae7d21-e7cd-446a-8fad-ce517018e348.png)

2. 2진수를 뒤에서 부터 6비트씩 끊어서 10진수로 변경한다.

![image](https://user-images.githubusercontent.com/80624082/135959183-cbabf2e0-cb31-4295-97d4-43af1bc04a20.png)

3. Base64 색인표를 보고 10진수에 해당하는 문자를 찾아낸 뒤 나열한다.

![image](https://user-images.githubusercontent.com/80624082/135959260-fbf7ba44-3735-41c8-9476-56a2d5c5f7c6.png)

연산이 막 복잡하지는 않습니다. 다들 시간 나실 때 한 번씩만 부탁드리겠습니다! 더 간단한 방법을 알고 계시는 분들은 코멘트 부탁드립니다!

1, 2 번은 진수 변환기를 쓰시면 편하실거고 [진수 변환기](http://www.hipenpal.com/tool/binary-octal-decimal-hexadecimal-number-converter-in-korean.php)
3번은 여기있는 Base64 색인표를 참고해주시면 됩니다. [Base64 색인표](https://ko.wikipedia.org/wiki/%EB%B2%A0%EC%9D%B4%EC%8A%A464)

Base32 테스트셋도 너무 잘만들어 주셨는데, 테스트 케이스를 받아올 때 '0x05'같은 한자리 16진수가 '0x0'과 '0x5'가 아닌, '0x5' 한자리로만 받아진다는 점을 명심해주시면 감사하겠습니다! 직접 테스트 케이스를 돌려보고 싶으신 분들은 아래 Discussion을 참고하시면 되겠습니다! 감사합니다!
https://github.com/euphony-io/check-in-listener/discussions/34#discussion-3604531

**Comment by kuro11pow2** [comment]
```cpp
TestParamType(std::vector<u_int8_t>{ 0xab, 0xcd }, "KvN")
```

**Comment by psh2849** [comment]
```C++
TestParamType(std::vector<u_int8_t>{ 0x3d, 0x2e }, "D0u")
```

**Comment by psh2849** [comment]
```c++
TestParamType(std::vector<u_int8_t>{ 0x4d, 0x5b }, "E1b")
```

**Comment by kuro11pow2** [comment]
0x1, 0x0, 0x0, 0x0, 0x0, 0x0 의 경우
1 0000 0000 0000 0000 0000 이니까
1 00000 00000 00000 00000 이렇게 되어야 하는데 그럼 BAAAA가 정답이 아닌가요?
PR하신 코드를 받아서 돌려보니 실패하네요 ㅠㅠㅠ

![image](https://user-images.githubusercontent.com/47289893/136031539-1d917052-1229-482a-aec0-070ad31db47d.png)

**Reply by YoungSeokHong** [reply]
앗 이거는 64진수(2의 6승)이여서 Base32처럼 5비트씩이 아니라 6비트씩 끊어주셔야 합니다!

**Reply by kuro11pow2** [reply]
아 ㅋㅋㅋㅋㅋㅋㅋㅋㅋㅋ 제가 착각했네요 그럼 0을 하나 더 붙여야 BAAAA가 나오는 걸까요?

**Reply by YoungSeokHong** [reply]
저도 이게 5비트 6비트 왔다갔다 하다보니 헷갈릴 때가 많더라구요ㅠㅠ
아까 코멘트는 정확하셨었는데 살짝 헷갈리셨었던 것 같습니다!

**Reply by YoungSeokHong** [reply]
네넵 그럴 것 같습니다!
같이 테스트셋 찾아주셔서 감사합니다!ㅠㅠ

**Comment by YoungSeokHong** [comment]
그리고 혹시나 해서 말씀드리는데, 제가 올려드린 [진수변환기](http://www.hipenpal.com/tool/binary-octal-decimal-hexadecimal-number-converter-in-korean.php)는 아쉽게도 앞의 '0'들은 생략해버립니다.
예를 들면 '0x3'이 '0011'이 아닌 그냥 '11'로 나와버립니다. 그래서 만약에 이 진수 변환기를 사용하실 때는 번거로우시더라도 글자 수 확인 부탁드립니다!
위에 두분은 제가 체크해봤는데, 너무 잘해주셔서 따로 확인하시지 않으셔도 될 것 같습니다!

**Comment by sujin9** [comment]
```c++
TestParamType(std::vector<u_int8_t>{ 0x9a, 0xc7 }, "JrH")
```

**Comment by dlwlstks96** [comment]
```Kotlin
TestParamType(std::vector<u_int8_t>{ 0x67, 0x6f, 0x6f, 0x64 }, "Bnb29k")
```

**Reply by YoungSeokHong** [reply]
32비트 짜리 테스트 케이스! 감사합니다!

**Comment by Jsueeee** [comment]
```cpp
TestParamType(std::vector<u_int8_t>{ 0xa8, 0xe9 }, "Kjp")
```

**Comment by eunzinn** [comment]
```c
TestParamType(std::vector<u_int8_t>{ 0xaa, 0xbb }, "Kq7")
```

**Comment by YoungSeokHong** [comment]
오 테스트 케이스가 엄청 많이 쌓였네요 ㅋㅋㅋ 감사합니다!
혹시 32비트 이상의 긴 테스트케이스도 도전해주실 분 계실까요??

**Comment by psh2849** [comment]
```c++
TestParamType(std::vector<u_int8_t>{ 0x6f, 0x3c, 0x4e, 0x5a, 0x6b, 0x2d, 0x1c, 0xaa }, "G88TlprLRyq")
```

**Reply by YoungSeokHong** [reply]
ㅋㅋㅋㅋㅋㅋㅋㅋㅋㅋㅋㅋㅋㅋㅋㅋ바로 올려주셨네요 감사합니다ㅠㅠ

**Comment by aiclaudev** [comment]
```cpp
TestParamType(std::vector<u_int8_t>{ 0x12, 0x34, 0x5a, 0x9f }, "ASNFqf")
```

**Comment by aiclaudev** [comment]
```cpp
TestParamType(std::vector<u_int8_t>{ 0x9f, 0x1b, 0x3c, 0x44, 0x55, 0x1d, 0x99, 0x80, 0x72, 0x17, 0x22, 0xcf  }, "nxs8RFUdmYByFyLP")
```

**Comment by ycs1m1yk** [comment]
```cpp
TestParamType(std::vector<u_int8_t>{ 0xc0, 0xde, 0xba, 0x5e, 0xba, 0x11 }, "MDeLpeLoR")
```

**Comment by YoungSeokHong** [comment]
@designe @yeonns 
멘토님! 테스트 케이스가 충분히 쌓인 것 같아 코멘트들 종합해서 base64Test.cpp에 추가한 뒤에 PR올리겠습니다!

**Reply by Jsueeee** [reply]
굿굿 👍👍

**Reply by designe** [reply]
좋아요!!

**Reply by YoungSeokHong** [reply]
테스트 케이스 올리려면 Base64부터 merge되어야하겠네요 ㅠㅠ Base32부터 올리겠습니다!
